src/DynamoCoreWpf/Utilities/NodeAutoCompleteUtilities.cs

```diff
@@ -0,0 +1,160 @@
+            foreach (var node in wsModel.Nodes)
+            {
+                if (nodesGuidsToConsider.Contains(node.GUID))
+                    continue;
+
+                if (connectedNodesBBox.IntersectsWith(node.Rect) ||
+                    connectedNodesBBox.Contains(node.Rect))
+                {
+                    intersectedNodes.Add(node);
+                    if (!realIntersection)
+                    {
+                        foreach (var connectedNode in nodesToConsider)
+                        {
+                            if (node.Rect.IntersectsWith(connectedNode.Rect) ||
+                                node.Rect.Contains(connectedNode.Rect) ||
+                                connectedNode.Rect.Contains(node.Rect))
+                            {
+                                realIntersection = true;
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return realIntersection;
+        }
+
+        /// <summary>
+        /// Automatically arranges misplaced nodes based on the specified parameters.
+        /// </summary>
+        /// <param name="wsModel">The workspace model containing the nodes to be arranged.</param>
+        /// <param name="queryNode">The node used as a starting point for the layout operation.</param>
+        /// <param name="misplacedNodes">A collection of nodes that are not properly positioned and need to be arranged.</param>
+        /// <param name="clusterLayout">Ensures misplaced nodes are positioned downstream. Nodes will be moved if necessary.</param>
+        /// <param name="checkWorkspaceNodes">Specifies whether to consider existing nodes in the workspace during the layout operation.</param>
+        /// <param name="finalizer">An action to be executed after the layout operation is complete, typically for cleanup or further adjustments.</param>
+        internal static void AutoLayoutNodes(WorkspaceModel wsModel,
+            NodeModel queryNode,
+            IEnumerable<NodeModel> misplacedNodes,
+            bool clusterLayout,
+            bool checkWorkspaceNodes,
+            Action finalizer)
+        {
+            DynamoSelection.Instance.Selection.AddRange(misplacedNodes);
+            wsModel.DoGraphAutoLayout(true, true, queryNode.GUID);
+
+            // For large clusters of nodes, auto-layout may place nodes on both sides of the query node.
+            // While the arrangement is technically fine, we move the entire group downstream for better consistency.
+            if (clusterLayout)
+            {
+                double offset = -1;
+                foreach (var node in misplacedNodes)
+                {
+                    if (node.X < queryNode.X)
+                    {
+                        offset = Math.Max(offset, queryNode.X - node.X);
+                    }
+                }
+
+                double balast = 50;
```

Comment:
Some of these constant numbers, we can define as class properties, but that can be new PRs